Microsoft.Toolkit.Uwp.UI.Controls/MosaicControl/MosaicControl.cs

```diff
@@ -0,0 +1,1098 @@
+
+                this.isImageSourceLoaded = false;
+
+                if (strategy == UIStrategy.Composition)
+                {
+                    var compositor = this.containerVisual.Compositor;
+
+                    using (var surfaceFactory = SurfaceFactory.CreateFromCompositor(compositor))
+                    {
+                        var surfaceUri = await surfaceFactory.CreateUriSurfaceAsync(uri);
+
+                        this.uriSurface = surfaceUri;
+                        this.brushVisual = compositor.CreateSurfaceBrush(surfaceUri.Surface);
+
+                        imageSize = surfaceUri.Size;
+                    }
+                }
+                else
+                {
+                    BitmapImage image = new BitmapImage();
+
+                    var storageFile = await StorageFile.GetFileFromApplicationUriAsync(uri);
+
+                    using (var stream = await storageFile.OpenReadAsync())
+                    {
+                        image.SetSource(stream);
+                    }
+
+                    this.brushXaml = new ImageBrush() { ImageSource = image };
+                    this.imageSize = new Size(image.PixelWidth, image.PixelHeight);
+                }
+
+                this.isImageSourceLoaded = true;
+
+                this.RefreshContainerMosaic();
+
+                this.RefreshImageSize(imageSize.Width, imageSize.Height);
+
+                if (isAnimated == true)
+                {
+                    this.IsAnimated = true;
+                }
+            }
+            finally
+            {
+                flag.Release();
+            }
+
+            if (this.ImageLoaded != null)
+            {
+                this.ImageLoaded(this, EventArgs.Empty);
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Gets or sets the scroll orientation of the mosaic.
+        /// Less images are drawn when you choose the Horizontal or Vertical value.
+        /// </summary>
+        public ScrollOrientation ScrollOrientation
+        {
+            get { return (ScrollOrientation)GetValue(ScrollOrientationProperty); }
+            set { SetValue(ScrollOrientationProperty, value); }
+        }
+
+        private static async void OnOrientationChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
+        {
+            var control = d as MosaicControl;
+            await control.RefreshContainerMosaicLocked();
+            await control.CreateModuloExpression(control.scrollviewer);
+        }
+
+        /// <inheritdoc/>
+        protected override async void OnApplyTemplate()
+        {
+            var strategy = this.Strategy;
+
+            // Gets the XAML root element
+            var rootElement = this.GetTemplateChild("RootElement") as FrameworkElement;
+
+            if (rootElement != null)
+            {
+                this.rootElement = rootElement;
+                rootElement.SizeChanged += RootElement_SizeChanged;
```

Comment:
`OnApplyTemplate()` can get called multiple times, so always safer to unhook before hooking any events in here,